Implement questions answered vs questions asked

cesswairimu opened the pull request:
Fixes #4137  


* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [x] code is in uniquely-named feature branch and has no merge conflicts
* [x] PR is descriptively titled
* [x] ask `@publiclab/reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software 

Thanks!

jywarren commented:
Hi! Just a hint - we can ignore the CodeClimate things, but sometimes they're really helpful suggestions. You can read them in detail and how to resolve them here: https://codeclimate.com/github/publiclab/plots2/pull/4256

jywarren commented on app/controllers/questions_controller.rb:
```diff
@@ -16,15 +16,42 @@ def filter_questions_by_tag(questions, tagnames)
   public
 
   def index
+    if params[:period]
```
This might be a good one to cache daily! I think esp. with collecting a year's worth of content each page load.

cesswairimu replied:
great! thanks

cesswairimu commented:
![past_year](https://user-images.githubusercontent.com/17081074/49895132-bf06a780-fe60-11e8-925e-8ed0ac6fa7f9.png)

cesswairimu commented:
Hey @gauravano  any suggestions on how I can best cache the stats here. Thanks.

gauravano commented:
As @jywarren mentioned we could cache for a day https://github.com/publiclab/plots2/pull/4256#discussion_r240290288 

we also cache home page data so you can see home_controller for reference. Let me know if you need help. Thanks!

cesswairimu commented:
Hey @jywarren I did some refactoring and removed the whole bunch of instance variables. Does the caching I did on the helper method make sense? Thanks

jywarren commented on app/helpers/questions_helper.rb:
```diff
@@ -0,0 +1,25 @@
+    SORTING_OPTIONS
+  end
+
+  def cache_quiz_stats
```
This works, but then people coding do have to remember to use the "cached version". Another way would be to build this inside the `filtering()` method itself, so all calls to that are filtered. If people really want the non-filtered version, they can manually flush it with... i forget but something like `Rails.cache.flush(...)`. Make sense? This is exciting!

cesswairimu replied:
Yes it does thanks, on it

cesswairimu commented:
@publiclab/reviewers this is ready kindly review

jywarren commented on app/helpers/questions_helper.rb:
```diff
@@ -0,0 +1,25 @@
+module QuestionsHelper
+  SORTING_OPTIONS = %w(All Week Month Year).freeze
+
+  def filtering(period)
```
This looks good but perhaps it could have a more descriptive name -- something describing how it's related to stats, or Q&A? Thanks!

jywarren replied:
Also, I'd love to see a simple test for these methods. Where would that live?

cesswairimu replied:
Let me try and figure out where the tests will go. Thanks

jywarren commented on app/helpers/questions_helper.rb:
```diff
@@ -0,0 +1,25 @@
+
+    if period == 'All'
+      Rails.cache.fetch("all_stats", expires_in: 1.days) do
+        @asked = Node.questions.to_a.size
```
I think if you use `Node.questions.count` it should do a more efficient query, just a SQL SIZE query instead of fetching all the full records and counting them. Maybe same on line below! 👍

cesswairimu replied:
@jywarren tried `.count` was not working fine  because its  a hash returning  results like so `{82=>1, 85=>2, 86=>2, 87=>1}`. Using `.length` instead. hope that's a better query?

jywarren commented on app/views/questions/index.html.erb:
```diff
@@ -1,4 +1,14 @@
+        </div>
+        <div class="col-md-4">
+            <%= form_tag request.url, :method => 'get'  do %>
+                <%= select_tag :period, options_for_select(options),onchange: "this.form.submit();", include_blank: "Filter stats", class: "form-control"%>
```
Ooh this is cool, could you share a screenshot? Thanks!

cesswairimu replied:
@jywarren a screenshot is :point_up: above. Thanks for the feedback.

jywarren commented:
Regarding the question of where statistics code "home base" is - would it fit in a model, so it's testable via unit tests? But open to ideas and brainstorming. Maybe ask @gauravano or @SidharthBansal for input on this as well!

cesswairimu commented:
![during](https://user-images.githubusercontent.com/17081074/50300646-1d9dd800-0496-11e9-9061-7ad638102c7a.png)
![month](https://user-images.githubusercontent.com/17081074/50300647-1e366e80-0496-11e9-9953-fac997a78379.png)


@jywarren I implemented the tests and changed the method name to a better name and improved the queries. Please take a look.

cesswairimu commented:
> Regarding the question of where statistics code "home base" is - would it fit in a model, so it's testable via unit tests? But open to ideas and brainstorming. Maybe ask @gauravano or @SidharthBansal for input on this as well!

@jywarren is this for this ticket #4317?

jywarren commented:
Actually it's more broad - that is, should we have a central location where we keep much of the statistics code, as if in a model, or should we be OK with some appearing in controllers, some in helpers... i'm just trying to think critically about how we're doing overall code organization. We don't have to come up with a final solution right now! Thanks!

jywarren commented:
This is great. Merging!!!
